src/helpers.js

```diff
@@ -155,3 +155,211 @@ export const isRectSelected = rect => selection => {
+        continue;
+      }
+      seen[cellPos] = true;
+
+      rowCells.push(cell);
+    }
+
+    rows.push(rowCells);
+  }
+
+  return rows;
+};
+
+// :: (tableNode: Node, tableArray: Array<Node>) -> Node
+// This function will transform a matrix of nodes
+// into table node respecting merged cells and rows configurations,
+// for example this array will be convert to the table below:
+//
+// array = [
+//   [A1, B1, C1, null],
+//   [A2, B2, null, D1],
+//   [A3. B3, C2, null],
+// ]
+//  ____________________________
+// |      |      |             |
+// |  A1  |  B1  |     C1      |
+// |______|______|______ ______|
+// |      |             |      |
+// |  A2  |     B2      |      |
+// |______|______ ______|      |
+// |      |      |      |  D1  |
+// |  A3  |  B3  |  C2  |      |
+// |______|______|______|______|
+//
+export const convertArrayOfRowsToTableNode = (tableNode, arrayOfNodes) => {
+  const rowsPM = [];
+  const map = TableMap.get(tableNode);
+  for (let rowIndex = 0; rowIndex < map.height; rowIndex++) {
+    const row = tableNode.child(rowIndex);
+    const rowCells = [];
+
+    for (let colIndex = 0; colIndex < map.width; colIndex++) {
+      if (!arrayOfNodes[rowIndex][colIndex]) {
+        continue;
+      }
+      const cellPos = map.map[rowIndex * map.width + colIndex];
+
+      const cell = arrayOfNodes[rowIndex][colIndex];
+      const cellData = [
+        Object.assign({}, cell.attrs),
+        cell.content,
+        cell.marks,
+      ];
+
+      const oldCell = tableNode.nodeAt(cellPos);
+      const newCell = oldCell.type.createChecked.apply(oldCell.type, cellData);
+      rowCells.push(newCell);
+    }
+
+    rowsPM.push(row.type.createChecked(row.attrs, rowCells, row.marks));
+  }
+
+  const newTable = tableNode.type.createChecked(
+    tableNode.attrs,
+    rowsPM,
+    tableNode.marks
+  );
+
+  return newTable;
+};
+
+export const moveTableColumn = (
+  table,
+  indexesOrigin,
+  indexesTarget,
+  direction
+) => {
+  let rows = transpose(convertTableNodeToArrayOfRows(table.node));
+
+  rows = moveRowInArrayOfRows(rows, indexesOrigin, indexesTarget, direction);
+  rows = transpose(rows);
+
+  return convertArrayOfRowsToTableNode(table.node, rows);
+};
+
+export const moveTableRow = (
+  table,
+  indexesOrigin,
+  indexesTarget,
+  direction
+) => {
+  let rows = convertTableNodeToArrayOfRows(table.node);
+
+  rows = moveRowInArrayOfRows(rows, indexesOrigin, indexesTarget, direction);
+
+  return convertArrayOfRowsToTableNode(table.node, rows);
+};
+
+const moveRowInArrayOfRows = (
+  rows,
+  indexesOrigin,
+  indexesTarget,
+  directionOverride
+) => {
+  let direction = indexesOrigin[0] > indexesTarget[0] ? -1 : 1;
+
+  const rowsExtracted = rows.splice(indexesOrigin[0], indexesOrigin.length);
+  const POSITION_OFFSET = rowsExtracted.length % 2 === 0 ? 1 : 0;
+  let target =
+    direction === -1
+      ? indexesTarget[0]
+      : indexesTarget[indexesTarget.length - 1] - POSITION_OFFSET;
```

Comment:
I would put this statement inside the `else` of the next block to avoid unnecessary computation.

Reply:
It looks like your `target` is just
```
const target =
    (direction === -1
      ? indexesTarget[0]
      : indexesTarget[indexesTarget.length - 1] - POSITION_OFFSET) + (directionOverride || 0);
```

Reply:
I liked your first suggestion I'll update the code.

the second one you are forgetting the override logic

Reply:
> the second one you are forgetting the override logic

What do you mean by that? You are just writing `indexesTarget[0]` and `indexesTarget[indexesTarget.length - 1] - POSITION_OFFSET` twice.

Reply:
No, I'm not, check this out:


```
// current one 
  if (directionOverride === -1 && direction === 1) {
    target = indexesTarget[0] - 1;
  } else if (directionOverride === 1 && direction === -1) {
    target = indexesTarget[indexesTarget.length - 1] - positionOffset + 1;
  } else {
    target =
      direction === -1
        ? indexesTarget[0]
        : indexesTarget[indexesTarget.length - 1] - positionOffset;
  }
```

```
// your suggestion
const target =
    (direction === -1
      ? indexesTarget[0]
      : indexesTarget[indexesTarget.length - 1] - POSITION_OFFSET) + (directionOverride || 0);

// same as above just more verbose
if (direction === -1) {
  // when the directions is -1 but we are overriding it with 1
  // we should use the other logic
  target = indexesTarget[0];
} else {
  // and here if we override it with -1
  // we should the logic above
  indexesTarget[indexesTarget.length - 1] - POSITION_OFFSET) + (directionOverride || 0);
}
```

Reply:
So, it's probably `if (directionOverride === -1 && direction === -1) ` and `if (directionOverride === 1 && direction === 1)` not `if (directionOverride === 1 && direction === -1) `.